src/coordinated_workers/coordinator.py

```diff
@@ -865,3 +911,352 @@ def _setup_charm_tracing(self):
+            urls.update({protocol: proxy_url})
+
+        return urls
+
+    @property
+    def _proxy_workload_tracing_receivers_urls(self) -> Dict[str, str]:
+        """Returns proxy workload tracing receivers urls published to the cluster."""
+        urls: Dict[str, str] = {}
+
+        for protocol in self._workload_tracing_receivers_urls:
+            scheme = "https" if self.tls_available else "http"
+            proxy_url = f"{scheme}://{self.hostname}:{self._proxy_worker_telemetry_port}{PROXY_WORKER_TELEMETRY_PATHS['workload_tracing_receivers_urls'].format(protocol=protocol)}"
+            urls.update({protocol: proxy_url})
+
+        return urls
+
+    def _validate_proxy_worker_telemetry_setup(
+        self,
+        workload_tracing_protocols: Union[List[ReceiverProtocol], None],  # type: ignore
+    ) -> None:
+        """Check if a valid proxy setup for worker telemetry is possible."""
+        # return true if proxying for worker telemetry is not enabled
+        if not self._proxy_worker_telemetry:
+            return
+        # if no workload protocol is defined, let the TracingEndpointRequirer handle this
+        if not workload_tracing_protocols:
+            return
+        # FIXME: GRPC should be allowed. Create an issue and link here.
+        for protocol in workload_tracing_protocols:  # type: ignore
+            if "grpc" in protocol:
+                logger.error(
+                    "Proxying worker telemetry via the coordinator failed."
+                    "Proxying worker telemetry does not support receiving workload traces via a GRPC based protocol."
+                )
+                break
+
+    def _setup_proxy_worker_telemetry(self) -> None:
+        """Extends nginx configuration with configurations required proxying worker telemetry to and from the workers via nginx."""
+        # Extend nginx config with worker metrics if enabled
+        worker_topology = self.cluster.gather_topology()
+        if worker_topology:
+            telemetry_upstreams, telemetry_locations = (
+                self._generate_worker_telemetry_nginx_config(worker_topology)
+            )
+            self._nginx_config.extend_upstream_configs(telemetry_upstreams)
+            self._nginx_config.update_server_ports_to_locations(
+                telemetry_locations, overwrite=False
+            )
+            self._update_upstreams_with_worker_telemetry_servers()
+
+    def _update_upstreams_with_worker_telemetry_servers(self) -> None:
+        """Update the upstreams in the nginx config to include the required servers/clients that send/receive worker telemetry."""
+        # Merge role-based and unit-based addresses collection for nginx config
+        # Every unit will get its own upstream for metric proxying
+        self._upstreams_to_addresses.update(self.cluster.gather_addresses_by_unit())
```

Comment:
this is a lot of mutation.
I think we should look for a way to pass the proxy configuration to whatever populates _upstreams_to_addresses instead of patching its outputs afterwards.